File: 백준/Gold/13460. 구슬 탈출 2/구슬 탈출 2.java
import java.util.*;
import java.io.*;

class Marble {
    int rx, ry;
    int bx, by;
    int cnt;

    public Marble (int rx, int ry, int bx, int by, int cnt) {
        this.rx = rx;
        this.ry = ry;
        this.bx = bx;
        this.by = by;
        this.cnt = cnt;
    }
}

/**
 * Tip
 * 2차원 배열을 회전시켜서(-90도) 생각해보면 
 * 배열의 1차원을 x로 둘 수 있다. 즉 map[0][1]이면 x=0, y=1이다. 
 */
public class Main {
    public static int N, M;
    public static char[][] map;
    public static boolean[][][][] visited;
    public static int holeX, holeY;
    public static Marble blue, red;

    public static int[] dx = {-1, 0, 1, 0};
    public static int[] dy = {0, 1, 0, -1};

    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    public static StringTokenizer st;

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        map = new char[N][M];
        visited = new boolean[N][M][N][M];
        
        for(int i = 0; i < N; i++) {
            String str = br.readLine();
            for(int j = 0; j < M; j++) {
                map[i][j] = str.charAt(j);

                if(map[i][j] == 'O') {
                    holeX = i;
                    holeY = j;
                } else if (map[i][j] == 'B') {
                    blue = new Marble(0, 0, i, j, 0);
                } else if (map[i][j] == 'R') {
                    red = new Marble(i, j, 0, 0, 0);
                }
            }
        }

        System.out.print(bfs());

        br.close();
    }

    public static int bfs() {
        Queue<Marble> queue = new ArrayDeque<>();
        queue.add(new Marble(red.rx, red.ry, blue.bx, blue.by, 1));
        visited[red.rx][red.ry][blue.bx][blue.by] = true;

        while(!queue.isEmpty()) {
            Marble marble = queue.poll();

            int curRx = marble.rx;
            int curRy = marble.ry;
            int curBx = marble.bx;
            int curBy = marble.by;
            int curCnt = marble.cnt;

            if(curCnt > 10) return -1;

            for(int i = 0; i < 4; i++) {
                int newRx = curRx;
                int newRy = curRy;
                int newBx = curBx;
                int newBy = curBy;

                boolean isRedHole = false;
                boolean isBlueHole = false;

                while(map[newRx + dx[i]][newRy + dy[i]] != '#') {
                    newRx += dx[i];
                    newRy += dy[i];

                    if(newRx == holeX && newRy == holeY) {
                        isRedHole = true;
                        break;
                    }
                }

                while(map[newBx + dx[i]][newBy + dy[i]] != '#') {
                    newBx += dx[i];
                    newBy += dy[i];

                    if(newBx == holeX && newBy == holeY) {
                        isBlueHole = true;
                        break;
                    }
                }

                // 너비우선탐색이므로 red만 홀에 들어간 경우 그 상황이 최소값이다.
                if(isBlueHole) continue;
                else if(isRedHole) {
                    return curCnt;
                }

                // 이동 중에 따지는 것 보다 다 이동시키고 다시 조정해주는 것이 낫다.
                if(newRx == newBx && newRy == newBy) {
                    if(i == 0) { // 위쪽으로 기울이기
                        // 더 큰 x값을 가지는 구슬이 뒤로 감
                        if(curRx > curBx) newRx -= dx[i];
                        else newBx -= dx[i];
                    } else if(i == 1) { // 오른쪽으로 기울이기
                        // 더 작은 y값을 가지는 구슬이 뒤로 감
                        if(curRy < curBy) newRy -= dy[i];
                        else newBy -= dy[i];
                    } else if(i == 2) { // 아래쪽으로 기울이기
                        // 더 작은 x값을 가지는 구슬이 뒤로 감
                        if(curRx < curBx) newRx -= dx[i];
                        else newBx -= dx[i];
                    } else { // 왼쪽으로 기울이기
                        // 더 큰 y값을 가지는 구슬이 뒤로 감
                        if(curRy > curBy) newRy -= dy[i];
                        else newBy -= dy[i];
                    }
                }

                // 두 구슬이 이동할 위치가 처음 방문하는 곳인 경우만 이동 -> 큐에 추가
                if(!visited[newRx][newRy][newBx][newBy]) {
                    visited[newRx][newRy][newBx][newBy] = true;
                    queue.add(new Marble(newRx, newRy, newBx, newBy, curCnt+1));
                }
            }
        }

        return -1;
    }
}

File: 백준/Gold/13460. 구슬 탈출 2/구슬 탈출 2.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import kotlin.collections.HashMap

const val RIGHT = 0
const val LEFT = 1
const val DOWN = 2
const val UP = 3
val dy = arrayOf(0, 0, 1, -1)
val dx = arrayOf(1, -1, 0, 0)
var n = 0
var m = 0
lateinit var origin: BoardInfo
lateinit var goal: Point

fun main() {
    initialize()
    val depth = bfs()
    println(if (depth in 1..10) depth else -1)
}

fun initialize() = with(BufferedReader(InputStreamReader(System.`in`))) {
    with(StringTokenizer(readLine())) {
        n = nextToken().toInt()
        m = nextToken().toInt()
    }

    var red: Point? = null
    var blue: Point? = null
    val map = Array(n) { y ->
        with(readLine()) {
            CharArray(m) { x ->
                this[x].also {
                    when (it) {
                        'R' -> red = Point(x, y)
                        'B' -> blue = Point(x, y)
                        'O' -> goal = Point(x ,y)
                    }
                }
            }
        }
    }
    origin = BoardInfo(map, red!!, blue!!)
    close()
}

fun bfs(): Int {
    val queue = LinkedList<BoardInfo>().apply {
        add(origin)
    }
    val visited = HashMap<String, Int>().apply {
        this[queue.peekFirst().toString()] = 0
    }

    while (queue.isNotEmpty()) {
        val current = queue.pollFirst()
        if (current.blue.x == goal.x && current.blue.y == goal.y) {
            continue
        } else if (current.red.x == goal.x && current.red.y == goal.y) {
            return visited[current.toString()]!!
        }

        for (direction in 0 until 4) {
            val next = current.move(direction)
            if (!visited.containsKey(next.toString())) {
                visited[next.toString()] = visited[current.toString()]!! + 1
                queue.addLast(next)
            }
        }
    }

    return -1
}

class BoardInfo(val map: Array<CharArray>, val red: Point, val blue: Point) {
    fun move(direction: Int): BoardInfo {
        val newMap = createNewMap()
        val newRed: Point
        val newBlue: Point
        if (isRedBeforeBlue(direction)) {
            newBlue = move(blue, direction, newMap)
            newRed = move(red, direction, newMap)
        } else {
            newRed = move(red, direction, newMap)
            newBlue = move(blue, direction, newMap)
        }
        return BoardInfo(newMap, newRed, newBlue)
    }

    private fun isRedBeforeBlue(direction: Int): Boolean {
        return (direction == RIGHT && red.x < blue.x) || (direction == LEFT && blue.x < red.x) || (direction == DOWN && red.y < blue.y) || (direction == UP && blue.y < red.y)
    }

    private fun createNewMap(): Array<CharArray> {
        return Array(n) {
            map[it].copyOf()
        }
    }

    private fun move(point: Point, direction: Int, newMap: Array<CharArray>): Point {
        val temp = newMap[point.y][point.x]
        newMap[point.y][point.x] = '.'
        var nx = point.x
        var ny = point.y
        do {
            if (nx == goal.x && ny == goal.y) {
                return Point(nx, ny)
            }
            nx += dx[direction]
            ny += dy[direction]
        } while (nx in 0 until m && ny in 0 until n && (newMap[ny][nx] == '.' || newMap[ny][nx] == 'O'))

        nx -= dx[direction]
        ny -= dy[direction]
        newMap[ny][nx] = temp
        return Point(nx, ny)
    }

    override fun toString(): String {
        return "${red.x},${red.y},${blue.x},${blue.y}"
    }
}

data class Point(val x: Int, val y: Int)
